source/server/listener_manager_impl.cc

```diff
@@ -205,27 +192,45 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
     ProtobufTypes::MessagePtr message =
         Config::Utility::translateToFactoryConfig(transport_socket, config_factory);
 
-    // Each transport socket factory owns one SslServerContext, we need to store them all in a
-    // vector since Ssl::ContextManager doesn't own SslServerContext. While transportSocketFacotry()
-    // always returns the first element of transport_socket_factories_, other transport socket
-    // factories are needed when the default Ssl::ServerContext updates SSL context based on
-    // ClientHello. This behavior is a workaround for initial SNI support before the full SNI based
-    // filter chain match is implemented.
-    transport_socket_factories_.emplace_back(config_factory.createTransportSocketFactory(
-        name_, sni_domains, skip_context_update, *message, *this));
-    ASSERT(transport_socket_factories_.back() != nullptr);
-  }
-  ASSERT(!transport_socket_factories_.empty());
-
-  // TODO(PiotrSikora): allow filter chains with mixed use of Session Ticket Keys.
-  // This doesn't work right now, because BoringSSL uses "session context" (initial SSL_CTX that
-  // accepted connection, before SNI update) for session related stuff, including Session Ticket
-  // callback, which is going to be called iff it's set on the initial SSL_CTX, even if it's not
-  // set on the current SSL_CTX that doesn't have any Session Ticket Keys configured.
-  if (has_stk != 0 && has_stk != has_tls) {
-    throw EnvoyException(fmt::format("error adding listener '{}': filter chains with mixed use of "
-                                     "Session Ticket Keys are currently not supported",
-                                     address_->asString()));
+    addFilterChain(
+        transport_socket.name(), sni_domains,
+        std::move(config_factory.createTransportSocketFactory(*message, *this, sni_domains)),
+        std::move(parent_.factory_.createNetworkFilterFactoryList(filter_chain.filters(), *this)));
+
+    // Check if downstream is expected to send data first (e.g. TLS, HTTP).
+    for (const auto& filter : filter_chain.filters()) {
+      if (filter.name() ==
+          Extensions::NetworkFilters::NetworkFilterNames::get().HTTP_CONNECTION_MANAGER) {
+        is_http = true;
+        break;
+      }
+    }
+
+    if (!downstream_sends_first) {
+      downstream_sends_first = is_tls || is_http || !sni_domains.empty();
+    } else if (downstream_sends_first.value() != (is_tls || is_http || !sni_domains.empty())) {
+      throw EnvoyException(fmt::format(fmt::format(
+          "error adding listener '{}': mixing incompatible filter chains", address_->asString())));
+    }
+
+    install_tls_inspector = is_tls || !sni_domains.empty();
+  }
+
+  // Automatically inject TLS Inspector if it wasn't configured explicitly and it's needed.
+  if (install_tls_inspector) {
```

Comment:
I actually don't like this code too much, and I agree that it should be explicitly configured, so I'm happy to remove it (which is one of the reasons why it was split into separate commit).

The downside of removing it, however, is that transport protocol and SNI-based filter chain matching won't work at all without it, and worse, traffic might be incorrectly routed to a filter chain with empty SNI and/or wrong transport protocol.

Reply:
With the current implementation of TLS sniffer, we have (and I should have caught this earlier, sorry!):

1. TLS sniffer: set `"tls"` or set `"raw_buffer"`.
2. 2nd sniffer: set `"dummy"` or set `"raw_buffer"`.

However, this clearly doesn't work with more than one sniffer, since 2nd sniffer would overwrite value set by TLS sniffer, therefore we need:

1. TLS sniffer: set `"tls"` or do nothing.
2. 2nd sniffer: set `"dummy"` or do nothing.
3. listener: set `"raw_buffer"` if nothing was detected.

See PR https://github.com/envoyproxy/envoy/pull/3260.

We could emit warning in 3., but after above change empty `detectedTransportProtocol()` is a valid state (i.e. no transport protocol was detected).

Also, I'd really like to be able to detect this at config time, and not at runtime, but nothing obvious comes to mind... Perhaps listener filters could indicate which transport protocols they can detect?

Reply:
(same is true for missing SNI, btw.)

Reply:
I think my general opinion here is to just inject it, as we do similar things in other places. If in the future someone really wants to not use TLS inspector we can make injection configurable. I would definitely add some type of warning though if we do inject telling people what we did, and make it more obvious if the config then fails to load because TLS inspector was not compiled/linked.

Reply:
I left the auto-inject, but removed "downstream send first" guard, since it wouldn't allow mixing with filter chains with custom transport sockets.